File: examples/shape_gallery.py
"""A gallery of `Path`-based shapes, laid out in a grid — a node defined
entirely in Python, the same way `tiled_layout.py` is.

Each shape is stored as plain data (an outline of anchors, a fill and a stroke
as RGB triples) rather than a live `dex.Path`: the geometry is just data, and
building the node at draw time keeps this gallery picklable. `draw` turns each
descriptor into a `dex.Path` and paints it into its cell, so every shape is
rendered by the real `Path` node — polygons, open polylines and Bézier curves
alike.
"""

import math

# Cell geometry, in node-local points.
COLS = 3
CELL_W = 250.0
CELL_H = 210.0
SHAPE_INSET = 30.0  # left padding of a shape within its cell
CAPTION_DROP = 150.0  # caption offset below the top of the cell


# -- shape descriptors ---------------------------------------------------
#
# A descriptor is `(name, fill, stroke, width, closed, anchors)` where:
#   * `fill`/`stroke` are RGB triples, or `None` for none;
#   * `closed` says whether the outline is filled and its ends joined;
#   * `anchors` is a list of `(x, y, in_handle, out_handle)`, each handle an
#     `(dx, dy)` offset from the point or `None` for a straight corner.


def _corners(points):
    """Anchors with no handles: a straight-edged outline through `points`."""
    return [(x, y, None, None) for (x, y) in points]


def _triangle():
    return (
        "triangle (polygon)",
        (80, 140, 240),
        (30, 60, 120),
        2.0,
        True,
        _corners([(60, 0), (120, 104), (0, 104)]),
    )


def _pentagon():
    r = 58.0
    points = [
        (r + r * math.cos(-math.pi / 2 + i * 2 * math.pi / 5),
         r + r * math.sin(-math.pi / 2 + i * 2 * math.pi / 5))
        for i in range(5)
    ]
    return ("pentagon (polygon)", (120, 200, 140), (40, 90, 60), 2.0, True, _corners(points))


def _star():
    outer, inner = 66.0, 27.0
    points = []
    for i in range(10):
        radius = outer if i % 2 == 0 else inner
        angle = -math.pi / 2 + i * math.pi / 5
        points.append((outer + radius * math.cos(angle), outer + radius * math.sin(angle)))
    return ("star (concave fill)", (250, 200, 60), (150, 110, 10), 2.0, True, _corners(points))


def _arrow():
    points = [(0, 30), (80, 30), (80, 0), (140, 55), (80, 110), (80, 80), (0, 80)]
    return ("arrow (polygon)", (230, 90, 90), None, 0.0, True, _corners(points))


def _wave():
    points = [(i * 4.5, 55.0 + 35.0 * math.sin(i * 4.5 / 22.0)) for i in range(31)]
    return ("wave (open polyline)", None, (70, 120, 220), 3.0, False, _corners(points))


def _blob():
    cx = cy = 62.0
    r = 56.0
    n = 6
    tangent = 2 * math.pi * r / n * 0.38
    anchors = []
    for i in range(n):
        angle = i * 2 * math.pi / n
        px, py = cx + r * math.cos(angle), cy + r * math.sin(angle)
        # A smooth vertex: the in handle mirrors the out handle.
        out = (-math.sin(angle) * tangent, math.cos(angle) * tangent)
        anchors.append((px, py, (-out[0], -out[1]), out))
    return ("blob (smooth anchors)", (180, 120, 230), (90, 50, 130), 2.0, True, anchors)


def _heart():
    # Four cubic segments with independent in/out handles.
    anchors = [
        (60, 108, (50, -38), (-50, -38)),
        (20, 20, (-20, 20), (15, -15)),
        (60, 28, (-5, -18), (5, -18)),
        (100, 20, (-15, -15), (20, 20)),
    ]
    return ("heart (bézier handles)", (230, 70, 100), (150, 30, 60), 2.0, True, anchors)


def _rounded_rect():
    # A rounded rectangle described purely as a Path: straight edges joined by
    # quarter-circle Bézier corners. This is `Rect`, expressed via `Path`.
    w, h, r = 150.0, 96.0, 22.0
    k = r * 0.5523  # cubic approximation of a quarter circle
    anchors = [
        (r, 0, (-k, 0), None),
        (w - r, 0, None, (k, 0)),
        (w, r, (0, -k), None),
        (w, h - r, None, (0, k)),
        (w - r, h, (k, 0), None),
        (r, h, None, (-k, 0)),
        (0, h - r, (0, k), None),
        (0, r, None, (0, -k)),
    ]
    return ("rounded rect (path arcs)", (90, 170, 210), (40, 90, 120), 2.0, True, anchors)


def default_shapes():
    """The built-in gallery, one descriptor per cell."""
    return [
        _triangle(), _pentagon(), _star(), _arrow(),
        _wave(), _blob(), _heart(), _rounded_rect(),
    ]


def _vec(x, y):
    return dex.Vector.new(float(x), float(y))


def build_path(spec):
    """Turn a descriptor into a live `dex.Path` node."""
    _name, fill, stroke, width, closed, anchors_data = spec

    anchors = []
    for (x, y, in_handle, out_handle) in anchors_data:
        anchor = dex.Anchor.corner(_vec(x, y))
        if in_handle is not None:
            anchor.in_handle = _vec(in_handle[0], in_handle[1])
        if out_handle is not None:
            anchor.out_handle = _vec(out_handle[0], out_handle[1])
        anchors.append(anchor)

    fill_color = dex.Color.rgb(*fill) if fill else dex.Color.transparent()
    stroke_style = dex.Stroke.new(width, dex.Color.rgb(*stroke)) if stroke else dex.Stroke.none()

    # `polygon` gives us a Path to populate; the fields carry the real outline.
    path = dex.Path.polygon([_vec(0, 0)], fill_color, stroke_style)
    path.anchors = anchors
    path.closed = closed
    return path


class ShapeGallery:
    def __init__(self, shapes=None):
        self.shapes = list(default_shapes() if shapes is None else shapes)

    # -- drawing ---------------------------------------------------------

    def draw(self, ctx):
        """Paint each shape into its grid cell and report the area taken."""
        origin = ctx.constraints.pos

        def box_at(x, y, w, h):
            return dex.DrawConstraints(
                pos=dex.ScreenPos.new(origin.x + x, origin.y + y),
                x=dex.AxisConstraint.Exactly(w),
                y=dex.AxisConstraint.Exactly(h),
                wrap=None,  # NotAllowed
                should_clip=False,
            )

        for i, spec in enumerate(self.shapes):
            col, row = i % COLS, i // COLS
            cell_x = col * CELL_W + SHAPE_INSET
            cell_y = row * CELL_H + 24.0
            ctx.draw_node(build_path(spec), box_at(cell_x, cell_y, CELL_W, CELL_H))
            ctx.draw_node(
                dex.Label.new(spec[0]),
                box_at(cell_x, cell_y + CAPTION_DROP, CELL_W, 20.0),
            )

        cols = min(len(self.shapes), COLS) or 1
        rows = (len(self.shapes) + COLS - 1) // COLS
        return dex.DrawResult.Complete(
            region=dex.ScreenRegion.from_min_size(
                origin,
                dex.Vector.new(cols * CELL_W + SHAPE_INSET, rows * CELL_H + 24.0),
            )
        )

    # -- lifecycle -------------------------------------------------------

    def type_name(self):
        return "Shape Gallery"


def transform():
    """A lambda returning the shape gallery."""
    return ShapeGallery()

File: sources/application/examples/shape_gallery.rs
//! A visual gallery of `Path`-based shapes, painted with the real
//! `Path::paint` renderer (concave fill, cubic-Bézier flattening).
//!
//! Run with: `cargo run -p dex --example shape_gallery`

use dex_core::prelude::*;
use dex_nodes::primitives::shapes::{Anchor, Path};
use eframe::egui;

/// Build an anchor at `(x, y)` with optional in/out handles (offsets from pos).
fn anchor(x: f32, y: f32, in_h: Option<(f32, f32)>, out_h: Option<(f32, f32)>) -> Anchor {
    Anchor {
        pos: Vector { x, y },
        in_handle: in_h.map(|(x, y)| Vector { x, y }),
        out_handle: out_h.map(|(x, y)| Vector { x, y }),
    }
}

fn v(x: f32, y: f32) -> Vector {
    Vector { x, y }
}

/// A triangle as a 3-corner closed polygon.
fn triangle() -> Path {
    Path::polygon(
        vec![v(60.0, 0.0), v(120.0, 104.0), v(0.0, 104.0)],
        Color::rgb(80, 140, 240),
        Stroke::new(2.0, Color::rgb(30, 60, 120)),
    )
}

/// A regular pentagon.
fn pentagon() -> Path {
    let r = 58.0;
    let points = (0..5)
        .map(|i| {
            let a = -std::f32::consts::FRAC_PI_2 + i as f32 * std::f32::consts::TAU / 5.0;
            v(r + r * a.cos(), r + r * a.sin())
        })
        .collect();
    Path::polygon(
        points,
        Color::rgb(120, 200, 140),
        Stroke::new(2.0, Color::rgb(40, 90, 60)),
    )
}

/// A five-point star: a concave polygon (exercises concave fill).
fn star() -> Path {
    let (outer, inner) = (66.0, 27.0);
    let points = (0..10)
        .map(|i| {
            let r = if i % 2 == 0 { outer } else { inner };
            let a = -std::f32::consts::FRAC_PI_2 + i as f32 * std::f32::consts::PI / 5.0;
            v(outer + r * a.cos(), outer + r * a.sin())
        })
        .collect();
    Path::polygon(
        points,
        Color::rgb(250, 200, 60),
        Stroke::new(2.0, Color::rgb(150, 110, 10)),
    )
}

/// A right-pointing block arrow.
fn arrow() -> Path {
    Path::polygon(
        vec![
            v(0.0, 30.0),
            v(80.0, 30.0),
            v(80.0, 0.0),
            v(140.0, 55.0),
            v(80.0, 110.0),
            v(80.0, 80.0),
            v(0.0, 80.0),
        ],
        Color::rgb(230, 90, 90),
        Stroke::NONE,
    )
}

/// An open sine wave as a polyline (no fill).
fn wave() -> Path {
    let points = (0..=30)
        .map(|i| {
            let x = i as f32 * 4.5;
            v(x, 55.0 + 35.0 * (x / 22.0).sin())
        })
        .collect();
    Path::polyline(points, Stroke::new(3.0, Color::rgb(70, 120, 220)))
}

/// A rounded blob: a closed path of smooth (mirrored-handle) anchors.
fn blob() -> Path {
    let (cx, cy, r, n) = (62.0, 62.0, 56.0, 6usize);
    let tangent = std::f32::consts::TAU * r / n as f32 * 0.38;
    let anchors = (0..n)
        .map(|i| {
            let a = i as f32 * std::f32::consts::TAU / n as f32;
            Anchor::smooth(
                v(cx + r * a.cos(), cy + r * a.sin()),
                v(-a.sin() * tangent, a.cos() * tangent),
            )
        })
        .collect();
    Path::closed_through(anchors, Color::rgb(180, 120, 230), Stroke::new(2.0, Color::rgb(90, 50, 130)))
}

/// A heart, from four cubic segments with independent in/out handles.
fn heart() -> Path {
    let anchors = vec![
        anchor(60.0, 108.0, Some((50.0, -38.0)), Some((-50.0, -38.0))),
        anchor(20.0, 20.0, Some((-20.0, 20.0)), Some((15.0, -15.0))),
        anchor(60.0, 28.0, Some((-5.0, -18.0)), Some((5.0, -18.0))),
        anchor(100.0, 20.0, Some((-15.0, -15.0)), Some((20.0, 20.0))),
    ];
    Path::closed_through(anchors, Color::rgb(230, 70, 100), Stroke::new(2.0, Color::rgb(150, 30, 60)))
}

/// A rounded rectangle built purely from a `Path`: straight edges joined by
/// quarter-circle Bézier corners. Demonstrates describing `Rect` via `Path`.
fn rounded_rect() -> Path {
    let (w, h, r) = (150.0, 96.0, 22.0);
    let k = r * 0.5523; // cubic approximation of a quarter circle
    let anchors = vec![
        anchor(r, 0.0, Some((-k, 0.0)), None),
        anchor(w - r, 0.0, None, Some((k, 0.0))),
        anchor(w, r, Some((0.0, -k)), None),
        anchor(w, h - r, None, Some((0.0, k))),
        anchor(w - r, h, Some((k, 0.0)), None),
        anchor(r, h, None, Some((-k, 0.0))),
        anchor(0.0, h - r, Some((0.0, k)), None),
        anchor(0.0, r, None, Some((0.0, -k))),
    ];
    Path::closed_through(anchors, Color::rgb(90, 170, 210), Stroke::new(2.0, Color::rgb(40, 90, 120)))
}

struct Gallery {
    shapes: Vec<(&'static str, Path)>,
}

impl eframe::App for Gallery {
    fn ui(&mut self, ui: &mut egui::Ui, _frame: &mut eframe::Frame) {
        let painter = ui.painter().clone();
        let origin = ui.max_rect().min;
        let (cols, cell_w, cell_h) = (3usize, 250.0_f32, 210.0_f32);
        for (i, (name, path)) in self.shapes.iter().enumerate() {
            let (col, row) = (i % cols, i / cols);
            let cell_x = origin.x + col as f32 * cell_w + 30.0;
            let cell_y = origin.y + row as f32 * cell_h + 24.0;
            path.paint(&painter, ScreenPos { x: cell_x, y: cell_y });
            painter.text(
                egui::pos2(cell_x, cell_y + cell_h - 74.0),
                egui::Align2::LEFT_TOP,
                *name,
                egui::FontId::proportional(14.0),
                egui::Color32::from_gray(60),
            );
        }
    }
}

fn main() -> eframe::Result {
    let shapes = vec![
        ("triangle (polygon)", triangle()),
        ("pentagon (polygon)", pentagon()),
        ("star (concave fill)", star()),
        ("arrow (polygon)", arrow()),
        ("wave (open polyline)", wave()),
        ("blob (smooth anchors)", blob()),
        ("heart (bézier handles)", heart()),
        ("rounded rect (path arcs)", rounded_rect()),
    ];
    let native_options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default().with_inner_size([780.0, 660.0]),
        ..Default::default()
    };
    eframe::run_native(
        "dex — shape gallery",
        native_options,
        Box::new(|cc| {
            cc.egui_ctx
                .options_mut(|opt| opt.theme_preference = egui::ThemePreference::Light);
            Ok(Box::new(Gallery { shapes }) as Box<dyn eframe::App>)
        }),
    )
}
